frontend/readhistory.lua

```diff
@@ -227,63 +212,64 @@ function ReadHistory:fileSettingsPurged(path)
 end
 
 function ReadHistory:clearMissing()
-    for i = #self.hist, 1, -1 do
-        if self.hist[i].file == nil or lfs.attributes(self.hist[i].file, "mode") ~= "file" then
-            self:removeItem(self.hist[i], i)
+    local history_updated
+    for i, v in ipairs(self.hist) do
+        if v.file == nil or lfs.attributes(v.file, "mode") ~= "file" then
+            self:removeItem(v, i, true) -- no flush
+            history_updated = true
         end
     end
-    self:ensureLastFile()
+    if history_updated then
+        self:_flush()
+        self:ensureLastFile()
+    end
 end
 
 function ReadHistory:removeItemByPath(path)
-    for i = #self.hist, 1, -1 do
-        if self.hist[i].file == path then
-            self:removeItem(self.hist[i])
-            break
-        end
+    local index = self:getIndexByFile(path)
+    if index then
+        self:removeItem(self.hist[index], index)
     end
-    self:ensureLastFile()
```

Comment:
Sure this one is not needed ?
(I guess it was me who added all these `self:ensureLastFile()`, may be too many just to be safe :)

Reply:
Yes, we do that in `removeItem`.

Reply:
OK.
(So, shouldn't you remove it also from inside ReadHistory:fileDeleted()? Or do it only in the dim = true branch if needed ?)